pkg/providers/gcp/provider_postgres.go

```diff
@@ -0,0 +1,693 @@
+		errMsg := fmt.Sprintf("failed to create or update secret %s, action was %s", sec.Name, result)
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrapf(err, errMsg)
+	}
+
+	gcpInstanceConfig, err := p.buildCloudSQLCreateStrategy(ctx, pg, strategyConfig, sec, address)
+	if err != nil {
+		msg := "failed to build and verify gcp cloudSQL instance configuration"
+		return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+
+	foundInstance, err := sqladminService.GetInstance(ctx, strategyConfig.ProjectID, gcpInstanceConfig.Name)
+	if err != nil && !resources.IsNotFoundError(err) {
+		msg := "cannot retrieve sql instance from gcp"
+		return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+	defer p.exposePostgresInstanceMetrics(ctx, pg, foundInstance)
+
+	if foundInstance != nil {
+		if !annotations.Has(pg, ResourceIdentifierAnnotation) {
+			annotations.Add(pg, ResourceIdentifierAnnotation, foundInstance.Name)
+			if err := p.Client.Update(ctx, pg); err != nil {
+				msg := "failed to add annotation to postgres cr"
+				return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+			}
+		}
+	}
+
+	if foundInstance == nil {
+		logger.Infof("no instance found, creating one")
+		_, err := sqladminService.CreateInstance(ctx, strategyConfig.ProjectID, gcpInstanceConfig.MapToGcpDatabaseInstance())
+		if err != nil && !resources.IsNotFoundError(err) {
+			msg := "failed to create cloudSQL instance"
+			return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+		}
+		annotations.Add(pg, ResourceIdentifierAnnotation, gcpInstanceConfig.Name)
+		if err := p.Client.Update(ctx, pg); err != nil {
+			msg := "failed to add annotation"
+			return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+		}
+		msg := "started cloudSQL provision"
+		return nil, croType.StatusMessage(msg), nil
+	}
+
+	if maintenanceWindowEnabled {
+		logger.Infof("building cloudSQL update config for: %s", foundInstance.Name)
+		modifiedInstance, err := p.buildCloudSQLUpdateStrategy(gcpInstanceConfig, foundInstance)
+		if err != nil {
+			msg := "error building update config for cloudsql instance"
+			return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+		}
+		if modifiedInstance != nil {
+			logger.Infof("modifying cloudSQL instance: %s", foundInstance.Name)
+			_, err := sqladminService.ModifyInstance(ctx, strategyConfig.ProjectID, foundInstance.Name, modifiedInstance)
+			if err != nil && !resources.IsConflictError(err) {
+				msg := fmt.Sprintf("failed to modify cloudsql instance: %s", foundInstance.Name)
+				return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+			}
+		}
+		_, err = controllerutil.CreateOrUpdate(ctx, p.Client, pg, func() error {
+			pg.Spec.MaintenanceWindow = false
+			return nil
+		})
+		if err != nil {
+			msg := "failed to set postgres maintenance window to false"
+			return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+		}
+	}
+
+	if foundInstance.State == "PENDING_CREATE" {
```

Comment:
That does seem logical, I'm unsure however if we can trigger an update on the instance early while it's `PENDING_CREATE` rather than waiting for the creation to be entirely completed to trigger the modify operation.

Reply:
If we're unsure then I would say better to move it above and wait. Or create a JIRA to test it.

Reply:
Addressed in 4a31dcd